File: tools/oracle/suites/balances.ts
// Oracle: account balances and net worth (src/domain/balances.ts, SPEC §6).
//
// Two separate questions, and conflating them is the classic bug:
//   1. WHAT IS THIS ACCOUNT WORTH?  — always opening + Σ its transactions.
//      Nothing ever changes this figure. Not archiving it, not excluding it.
//   2. WHAT GOES INTO THE TOTAL?    — countsTowardNetWorth(): not archived AND
//      not flagged excludeFromNetWorth.
// A port that implements exclusion by zeroing or hiding the balance passes
// every net-worth case here and fails every accountBalances one.
//
// Pending transactions COUNT in balanceMinor (D15) and are the difference
// between it and clearedMinor. A missing rate removes a currency from the
// TOTAL and is named in missingRateCurrencies — the money is never guessed at
// and never silently dropped.
import { accountBalances, balanceFromAmounts, countsTowardNetWorth, netWorth } from '../../../src/domain/balances';
import type { Account } from '../../../src/db/types';
import { loadBook, materialiseBook, type Book } from '../book';
import { EXCLUSIONS_BOOK, EXCLUSIONS_SIMPLE_BOOK, GOLDEN_BOOK, ROLLUP_BOOK } from '../books';
import { Cases, ORACLE_VERSION, type OracleFile } from '../oracle';

/**
 * Project AccountBalance onto the fields a re-implementation must reproduce.
 * The full record carries the whole Account row (colour, groupId, …), which is
 * storage detail, not arithmetic — putting it in the fixture would demand a
 * Swift port have columns it may not need.
 */
async function balanceRows(): Promise<unknown[]> {
  return (await accountBalances()).map((b) => ({
    accountId: b.account.id,
    name: b.account.name,
    currency: b.account.currency,
    balanceMinor: b.balanceMinor,
    clearedMinor: b.clearedMinor,
    txCount: b.txCount,
    excludedFromNetWorth: b.excludedFromNetWorth,
  }));
}

const netWorthShape = async (): Promise<unknown> => {
  const nw = await netWorth();
  return {
    totalBaseMinor: nw.totalBaseMinor,
    baseCurrency: nw.baseCurrency,
    missingRateCurrencies: nw.missingRateCurrencies,
    excludedCount: nw.excludedCount,
    excludedBaseMinor: nw.excludedBaseMinor,
  };
};

const account = (over: Partial<Account>): Account => ({
  id: 'a', name: 'A', type: 'current', currency: 'GBP', openingBalanceMinor: 0,
  colour: '#000', groupId: null, sortOrder: 0, archived: false, ...over,
});

export async function balancesSuite(): Promise<OracleFile> {
  const c = new Cases();
  const books: Record<string, Book> = {
    golden: materialiseBook(GOLDEN_BOOK),
    exclusions: materialiseBook(EXCLUSIONS_BOOK),
    'exclusions-simple': materialiseBook(EXCLUSIONS_SIMPLE_BOOK),
    rollup: materialiseBook(ROLLUP_BOOK),
  };

  // ------------------------------------------------------------ pure core
  const sums: [string, number, number[], number, string][] = [
    ['plain', 100_000, [-4_567, -5_433], 90_000, 'opening plus the signed amounts, nothing else'],
    ['empty', 50_000, [], 50_000, 'an account with no transactions is worth its opening balance'],
    ['negative', 0, [-1_000, -2_000], -3_000, 'a balance may be negative (a credit card)'],
    ['back-to-zero', 1_000, [-1_000], 0, 'exact cancellation is exactly zero'],
  ];
  for (const [slug, opening, amounts, expected, describes] of sums) {
    c.hand(
      `balances.sum.${slug}`, describes, 'balances.balanceFromAmounts',
      { openingMinor: opening, amounts },
      { value: balanceFromAmounts(opening, amounts) },
      { value: expected },
    );
  }

  const counts: [string, boolean, boolean, boolean, string][] = [
    ['plain', false, false, true, 'an ordinary account counts'],
    ['archived', true, false, false, 'an archived (retired) account does not'],
    ['excluded', false, true, false, 'a “show it, don’t count it” account does not'],
    ['both', true, true, false, 'the two reasons compose — either one is enough to leave it out'],
  ];
  for (const [slug, archived, excluded, expected, describes] of counts) {
    c.hand(
      `balances.counts.${slug}`, describes, 'balances.countsTowardNetWorth',
      { account: { archived, excludeFromNetWorth: excluded } },
      { value: countsTowardNetWorth(account({ archived, excludeFromNetWorth: excluded })) },
      { value: expected },
    );
  }
  c.hand(
    'balances.counts.flag-absent',
    'a row written before the exclusion flag existed (the key is simply absent) counts — undefined resolves to false',
    'balances.countsTowardNetWorth',
    { account: { archived: false } },
    { value: countsTowardNetWorth(account({})) },
    { value: true },
  );

  // -------------------------------------------------------- golden month
  // Hand calculation, from the comment block of tests/golden.test.ts:
  //   Current = 100000 +250000 -4567 -5433 -10000 +1000 -20000 -3000 = 308000
  //   Savings = 50000 +20000                                         =  70000
  //   Holiday = 20000 -2000                                          =  18000 (EUR)
  //   Net worth = 308000 + 70000 + (18000 × 0.85 = 15300)            = 393300
  await loadBook(books.golden);
  c.hand(
    'balances.golden.accounts',
    'the golden month’s three account balances, hand-calculated: £3,080.00 current (the £30 pending included), £700.00 savings, €180.00 holiday in its OWN currency',
    'balances.accountBalances',
    { book: 'golden' },
    { rows: await balanceRows() },
    {
      rows: [
        { accountId: 'cur', name: 'Current', currency: 'GBP', balanceMinor: 308_000, clearedMinor: 311_000, txCount: 7, excludedFromNetWorth: false },
        { accountId: 'sav', name: 'Savings', currency: 'GBP', balanceMinor: 70_000, clearedMinor: 70_000, txCount: 1, excludedFromNetWorth: false },
        { accountId: 'hol', name: 'Holiday', currency: 'EUR', balanceMinor: 18_000, clearedMinor: 18_000, txCount: 1, excludedFromNetWorth: false },
      ],
    },
    { carriedFrom: 'tests/golden.test.ts', note: 'clearedMinor excludes the £30.00 pending Uber: 308000 + 3000 = 311000. A balance is NEVER converted — the holiday account is €180.00.' },
  );
  c.hand(
    'balances.golden.net-worth',
    'net worth converts the EUR account once, at 0.85, and totals £3,933.00',
    'balances.netWorth',
    { book: 'golden' },
    await netWorthShape(),
    {
      totalBaseMinor: 393_300,
      baseCurrency: 'GBP',
      missingRateCurrencies: [],
      excludedCount: 0,
      excludedBaseMinor: 0,
    },
    { carriedFrom: 'tests/golden.test.ts' },
  );

  // ---------------------------------------------------------- exclusions
  await loadBook(books.exclusions);
  c.hand(
    'balances.exclusions.accounts',
    'every account keeps its own real balance — archiving and excluding change what the TOTAL counts, never what an account is worth',
    'balances.accountBalances',
    { book: 'exclusions' },
    { rows: await balanceRows() },
    {
      rows: [
        { accountId: 'a-main', name: 'Main', currency: 'GBP', balanceMinor: 97_500, clearedMinor: 100_000, txCount: 1, excludedFromNetWorth: false },
        { accountId: 'a-arch', name: 'Old ISA', currency: 'GBP', balanceMinor: 499_000, clearedMinor: 499_000, txCount: 1, excludedFromNetWorth: false },
        { accountId: 'a-excl', name: 'Gift Cards', currency: 'GBP', balanceMinor: 21_000, clearedMinor: 21_000, txCount: 1, excludedFromNetWorth: true },
        { accountId: 'a-usd', name: 'Dollar Pot', currency: 'USD', balanceMinor: 10_000, clearedMinor: 10_000, txCount: 0, excludedFromNetWorth: false },
        { accountId: 'a-chf', name: 'Swiss Pot', currency: 'CHF', balanceMinor: 20_000, clearedMinor: 20_000, txCount: 0, excludedFromNetWorth: false },
        { accountId: 'a-excl-chf', name: 'Lent to Ana', currency: 'CHF', balanceMinor: 5_000, clearedMinor: 5_000, txCount: 0, excludedFromNetWorth: true },
      ],
    },
    { note: 'Rows come back in sortOrder, then name. Main is £1,000.00 minus a £25.00 PENDING charge, so balanceMinor and clearedMinor differ.' },
  );
  c.hand(
    'balances.exclusions.net-worth',
    'the total is £975.00 + $100.00×0.79 = £1,054.00; CHF is named as missing rather than dropped in silence, and the “not counted” total REFUSES to be a number because one excluded account cannot be converted',
    'balances.netWorth',
    { book: 'exclusions' },
    await netWorthShape(),
    {
      totalBaseMinor: 105_400,
      baseCurrency: 'GBP',
      missingRateCurrencies: ['CHF'],
      excludedCount: 2,
      excludedBaseMinor: null,
    },
    { note: 'excludedBaseMinor null is the honest answer, not an error: a partial “not counted” figure would silently omit an account. Archived accounts are not in excludedCount even when flagged — they are already out for an older reason.' },
  );

  await loadBook(books['exclusions-simple']);
  c.hand(
    'balances.exclusions-simple.net-worth',
    'with every excluded account convertible, the “not counted” figure is a real number: £250.00 sitting outside a £1,000.00 total',
    'balances.netWorth',
    { book: 'exclusions-simple' },
    await netWorthShape(),
    {
      totalBaseMinor: 100_000,
      baseCurrency: 'GBP',
      missingRateCurrencies: [],
      excludedCount: 1,
      excludedBaseMinor: 25_000,
    },
  );

  // ------------------------------------------------------- wider scenario
  await loadBook(books.rollup);
  c.derived(
    'balances.rollup.accounts',
    'the reports book’s balances: both legs of a transfer move real money, and a transaction in an unrated currency still changes its own account’s balance',
    'balances.accountBalances',
    { book: 'rollup' },
    { rows: await balanceRows() },
  );
  c.derived(
    'balances.rollup.net-worth',
    'net worth over that book: the CHF account has no rate, so CHF is named and its money is left out of the total entirely',
    'balances.netWorth',
    { book: 'rollup' },
    await netWorthShape(),
  );

  return {
    oracleVersion: ORACLE_VERSION,
    area: 'balances',
    title: 'Balances: per-account, cleared vs pending, and net worth with exclusions',
    generatedFrom: ['src/domain/balances.ts'],
    notes: [
      'balanceMinor = openingBalanceMinor + Σ(all that account’s transactions), pending included (D15).',
      'clearedMinor is the same sum over cleared transactions only.',
      'An account balance is always in the ACCOUNT’s currency and is never converted.',
      'Net worth counts an account iff it is neither archived nor flagged excludeFromNetWorth; each counted account is converted to base ONCE.',
      'A currency with no rate to base is named in missingRateCurrencies and contributes nothing — it is never approximated.',
      'excludedBaseMinor is null when any excluded account cannot be converted: an incomplete “not counted” total would be a wrong number.',
      'balances.accountBalances rows are ordered by sortOrder, then by name.',
    ],
    books,
    cases: c.list,
  };
}

File: native/Sources/MyMoneyKit/Domain/Balances.swift
// Account balances and net worth, ported from src/domain/balances.ts.
//
// A BALANCE IS ALWAYS THE ACCOUNT'S OWN REAL BALANCE, in the account's own
// currency, and it is never converted. Whether an account lands in the NET
// WORTH TOTAL is a completely separate question, answered by
// `countsTowardNetWorth`. Archived (retired) or `excludeFromNetWorth`
// ("show it, don't count it") keep an account out of the TOTAL while leaving
// its balance, its transactions and its visibility untouched. Conflating the
// two -- zeroing an excluded account's balance, or hiding it -- would mean the
// owner could not find their own money, which is worse than a wrong headline.
//
// PURE FUNCTIONS OVER ARRAYS, no database. The TypeScript streams the
// transactions table in batches because the sidebar re-runs it on every write
// at 100,000 rows; that is a storage concern and belongs with whatever
// persistence layer the app grows (SQLite, and it will express this as a
// GROUP BY). What belongs HERE is the arithmetic, in a form a test can hand a
// handful of rows to and a form the oracle's books can drive directly.
import Foundation

public struct AccountBalance: Sendable, Hashable {
    public let account: Account
    /// Opening balance plus ALL transactions, pending included (D15).
    public let balanceMinor: Int64
    /// Opening balance plus CLEARED transactions only.
    public let clearedMinor: Int64
    public let txCount: Int
    /// Mirror of `account.excludeFromNetWorth`, resolved here so a consumer can
    /// render "not counted" without a second lookup. The balance above is the
    /// real balance either way.
    public let excludedFromNetWorth: Bool
}

public struct NetWorth: Sendable, Hashable {
    /// The balances that COUNT -- not archived, not excluded -- totalled IN
    /// THEIR OWN CURRENCY and each currency's subtotal then converted to base
    /// exactly once. Per currency, not per account: see `netWorth` below for
    /// why the difference is a real penny and not a stylistic one.
    public let totalBaseMinor: Int64
    public let baseCurrency: String
    /// Currencies left OUT of the total because no rate to base exists.
    /// Surfaced, never guessed: the honest total is the one that says what is
    /// missing from it. In the order the accounts were encountered, matching
    /// the TypeScript's insertion-ordered Set -- note that the BACKUP MANIFEST
    /// sorts the same list instead (see Manifest.swift), and the difference is
    /// deliberate in both places.
    public let missingRateCurrencies: [String]
    /// How many VISIBLE (non-archived) accounts are flagged not-counted, so the
    /// UI can say "N accounts not counted" honestly. An archived account that
    /// is also flagged is not counted here: it is already out of the total for
    /// an older, separate reason and is not on screen next to the figure.
    public let excludedCount: Int
    /// What those excluded accounts are worth in base currency -- totalled per
    /// currency and converted once, exactly like the figure above.
    ///
    /// nil the moment ONE of them cannot be converted. A partial "not counted"
    /// figure would silently omit an account, and a number that quietly means
    /// something other than what it says is the failure this whole package
    /// exists to prevent. Their currencies are deliberately NOT added to
    /// `missingRateCurrencies`: that list means "your total is missing this
    /// currency", and an excluded account was never going to be in the total.
    public let excludedBaseMinor: Int64?
}

public enum Balances {
    /// Has the owner flagged this account out of net-worth totals?
    /// `== true`, never a truthiness test and never `?? false` spelled some
    /// other way: this is the TypeScript's `account.excludeFromNetWorth ===
    /// true`. Absent and an explicit `false` are different bytes in the file
    /// (see `Account.excludeFromNetWorth`) and the same answer here, which is
    /// the whole point -- a net worth that changed because a row was written by
    /// an older build would be indefensible.
    public static func isExcludedFromNetWorth(_ account: Account) -> Bool {
        account.excludeFromNetWorth == true
    }

    /// Does this account contribute to the net-worth total?
    ///
    /// The two reasons not to count COMPOSE: archived OR excluded implies not
    /// counted. One function, so the headline figure and any chart drawn later
    /// can never disagree about which accounts count.
    public static func countsTowardNetWorth(_ account: Account) -> Bool {
        !account.archived && !isExcludedFromNetWorth(account)
    }

    /// The pure core: opening + the sum of the amounts.
    public static func balanceFromAmounts(_ openingMinor: Int64, _ amounts: [Int64]) throws -> Int64 {
        try Money.sum(amounts, startingAt: openingMinor)
    }

    struct Aggregate {
        var sum: Int64 = 0
        var cleared: Int64 = 0
        var count: Int = 0
    }

    /// accountId -> totals, in one pass.
    static func aggregate(_ transactions: [Transaction]) throws -> [String: Aggregate] {
        var totals: [String: Aggregate] = [:]
        for tx in transactions {
            var entry = totals[tx.accountId] ?? Aggregate()
            let (sum, sumOverflowed) = entry.sum.addingReportingOverflow(tx.amountMinor)
            if sumOverflowed {
                throw MoneyError.overflow("summing transactions for account \(tx.accountId)")
            }
            entry.sum = sum
            if tx.status == .cleared {
                let (cleared, clearedOverflowed) = entry.cleared.addingReportingOverflow(tx.amountMinor)
                if clearedOverflowed {
                    throw MoneyError.overflow("summing cleared transactions for account \(tx.accountId)")
                }
                entry.cleared = cleared
            }
            entry.count += 1
            totals[tx.accountId] = entry
        }
        return totals
    }

    /// Balances for ALL accounts -- archived and net-worth-excluded included.
    /// Callers filter; an excluded account must stay visible with its real
    /// balance.
    ///
    /// Sorted by `sortOrder`, then by name. Two notes on the ordering, because
    /// this is a place a port can silently diverge:
    ///  * the sort is made STABLE by falling back to the input position.
    ///    JavaScript's `Array.prototype.sort` has been required to be stable
    ///    since ES2019 and Swift's `sorted(by:)` is not, so two accounts
    ///    identical in sortOrder AND name would otherwise come back in an order
    ///    that depended on the standard library's mood;
    ///  * the name comparison is `localeCompare`'s, i.e. locale-aware, pinned
    ///    to en_GB here. The TypeScript passes no locale and gets the browser's.
    ///    Nothing in the oracle exercises a name tie, so this is a judgement,
    ///    and it is recorded as one.
    public static func accountBalances(
        accounts: [Account],
        transactions: [Transaction]
    ) throws -> [AccountBalance] {
        let totals = try aggregate(transactions)
        let enGB = Locale(identifier: "en_GB")
        let rows: [(offset: Int, balance: AccountBalance)] = try accounts.enumerated().map { index, account in
            let entry = totals[account.id] ?? Aggregate()
            let (balance, balanceOverflowed) = account.openingBalanceMinor.addingReportingOverflow(entry.sum)
            let (cleared, clearedOverflowed) = account.openingBalanceMinor.addingReportingOverflow(entry.cleared)
            if balanceOverflowed || clearedOverflowed {
                throw MoneyError.overflow("balance of account \(account.id)")
            }
            return (
                index,
                AccountBalance(
                    account: account,
                    balanceMinor: balance,
                    clearedMinor: cleared,
                    txCount: entry.count,
                    excludedFromNetWorth: isExcludedFromNetWorth(account)
                )
            )
        }
        return rows.sorted { lhs, rhs in
            let a = lhs.balance.account
            let b = rhs.balance.account
            if a.sortOrder != b.sortOrder { return a.sortOrder < b.sortOrder }
            let byName = a.name.compare(b.name, options: [], range: nil, locale: enGB)
            if byName != .orderedSame { return byName == .orderedAscending }
            return lhs.offset < rhs.offset
        }.map(\.balance)
    }

    /// Net worth in base currency.
    ///
    /// SUM PER CURRENCY FIRST, CONVERT ONCE -- matching `Reports.netWorthSeries`
    /// and `Manifest.compute`'s current rule (`.perCurrency`) exactly.
    ///
    /// This used to convert once per ACCOUNT, and that made the headline figure
    /// disagree with the right-hand end of the net-worth chart whenever two
    /// counted accounts shared a non-base currency: two EUR 7.05 accounts at
    /// 0.85 round to 599 + 599 = 1198 per account, but 1410 x 0.85 = 1198.5 ->
    /// 1199 per currency. Both are defensible in isolation; showing BOTH, in
    /// two places, for the same book, is not.
    ///
    /// Per currency is the one to keep: it rounds once instead of once per
    /// account, so the error cannot grow with the number of accounts, and it is
    /// the ordinary accounting treatment -- total in the source currency, then
    /// convert. It is also what the chart already did, so the chart's history
    /// stays truthful rather than being retroactively re-rounded.
    ///
    /// A balance is still never converted for its own sake: conversion happens
    /// HERE and only here, on a TOTAL, rounded half away from zero exactly
    /// once. Not per transaction (which would round thousands of times and
    /// drift), and never on a sum of mixed currencies (which is not a number).
    ///
    /// THE "NOT COUNTED" TOTAL FOLLOWS THE SAME RULE, for the same reason: it
    /// is a total, it sits on screen beside the headline, and two totals on one
    /// screen rounded two different ways is the defect this change exists to
    /// remove.
    ///
    /// Ported from src/domain/balances.ts `netWorth()`; the two must move
    /// together, and the manifest's `.perCurrency` branch with them.
    public static func netWorth(
        _ balances: [AccountBalance],
        baseCurrency: String,
        rates: RateTable
    ) throws -> NetWorth {
        var missing: [String] = []
        var excludedCount = 0
        // Dictionaries plus the order their keys were first seen: the
        // TypeScript accumulates into Maps, which iterate in insertion order,
        // and `missingRateCurrencies` is documented to be in encounter order.
        // A Swift Dictionary iterates in whatever order it likes, so the order
        // is carried explicitly rather than hoped for.
        var countedOrder: [String] = []
        var counted: [String: Int64] = [:]
        var excludedOrder: [String] = []
        var excluded: [String: Int64] = [:]

        for row in balances {
            if row.account.archived { continue }
            let currency = row.account.currency
            if row.excludedFromNetWorth {
                excludedCount += 1
                if excluded[currency] == nil {
                    excluded[currency] = 0
                    excludedOrder.append(currency)
                }
                let (next, overflowed) = excluded[currency]!.addingReportingOverflow(row.balanceMinor)
                if overflowed { throw MoneyError.overflow("excluded balances in \(currency)") }
                excluded[currency] = next
                continue
            }
            if counted[currency] == nil {
                counted[currency] = 0
                countedOrder.append(currency)
            }
            let (next, overflowed) = counted[currency]!.addingReportingOverflow(row.balanceMinor)
            if overflowed { throw MoneyError.overflow("counted balances in \(currency)") }
            counted[currency] = next
        }

        var total: Int64 = 0
        for currency in countedOrder {
            switch Money.convert(minor: counted[currency]!, from: currency, to: baseCurrency, using: rates) {
            case .converted(let value):
                let (next, overflowed) = total.addingReportingOverflow(value)
                if overflowed { throw MoneyError.overflow("net worth total") }
                total = next
            case .missingRate:
                // Insertion order, and de-duplicated by construction: the
                // TypeScript spreads a Set, which in JavaScript iterates in
                // insertion order.
                missing.append(currency)
            case .notRepresentable:
                // NOT folded into "no rate": there IS a rate, and telling the
                // owner otherwise would send them to the rates screen to fix
                // something that is not broken.
                throw MoneyError.notRepresentable("the \(currency) subtotal")
            }
        }

        // Starts at 0 and LATCHES to nil the moment one excluded currency
        // cannot be converted -- a partial "not counted" total would be a wrong
        // number, and a wrong number is worse than an honest gap (SPEC 6).
        var excludedBase: Int64? = 0
        for currency in excludedOrder {
            switch Money.convert(minor: excluded[currency]!, from: currency, to: baseCurrency, using: rates) {
            case .converted(let value):
                if let running = excludedBase {
                    let (next, overflowed) = running.addingReportingOverflow(value)
                    if overflowed { throw MoneyError.overflow("total of excluded accounts") }
                    excludedBase = next
                }
            case .missingRate:
                excludedBase = nil
            case .notRepresentable:
                throw MoneyError.notRepresentable("the excluded \(currency) subtotal")
            }
        }

        return NetWorth(
            totalBaseMinor: total,
            baseCurrency: baseCurrency,
            missingRateCurrencies: missing,
            excludedCount: excludedCount,
            excludedBaseMinor: excludedBase
        )
    }
}
